commatrix/generate.go

```diff
@@ -0,0 +1,175 @@
+package commatrix
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"sync"
+
+	"golang.org/x/sync/errgroup"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	clientutil "github.com/openshift-kni/commatrix/client"
+	"github.com/openshift-kni/commatrix/consts"
+	"github.com/openshift-kni/commatrix/debug"
+	"github.com/openshift-kni/commatrix/ss"
+	"github.com/openshift-kni/commatrix/types"
+)
+
+// generate relevant matrix and diff files
+func GenerateCommatrix(kubeconfig, customEntriesPath, customEntriesFormat, format string, env Env, deployment Deployment, printFn func(m types.ComMatrix) ([]byte, error), destDir string) {
```

Comment:
this should not receive a printFn (this is a pure implementation detail, the caller does not need to know that it exists) since the "format" is enough.
Also, it seems that the function does a lot more than just generating a matrix (generates, writes to file, diffs, more writes, etc.), I'd say it is worth splitting these functionalities to different functions and letting the caller decide what it wants (i.e this function should only return a matrix).